File: Leetcode-Problems/Asteroid Collisions.cpp
class Solution
{
public:
    vector<int> asteroidCollision(vector<int> &asteroids)
    {
        vector<int> st;
        for (int i : asteroids)
        {
            bool flag = true;
            while (!st.empty() && i < 0 && st.back() > 0)
            {
                if (abs(i) > st.back())
                    st.pop_back();
                else if (abs(i) == st.back())
                {
                    st.pop_back();
                    flag = false;
                    break;
                }
                else
                {
                    flag = false;
                    break;
                }
            }
            if (flag)
                st.push_back(i);
        }
        return st;
    }
};

File: Leetcode-Problems/Asteroid Collisions.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> st = new Stack<>();
        for (int i : asteroids) {
            while (!st.isEmpty() && i < 0 && st.peek() > 0) {
                int diff = i + st.peek();
                if (diff < 0) {
                    st.pop();
                } else if (diff > 0) {
                    i = 0;
                } else {
                    i = 0;
                    st.pop();
                }
            }
            if (i != 0)
                st.push(i);
        }
        int[] arr = new int[st.size()];
        int i = st.size() - 1;
        while (!st.isEmpty())
            arr[i--] = st.pop();
        return arr;
    }
}